File: strings/07_roman_to_int/roman_to_int.cpp
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <stdexcept>
#include <cstdint>

using namespace std;


class Solution {
public:
    int romanToInt(const string& s) {
        if (!checkConstraints(s)) {
            throw std::runtime_error("Constraints violated");
        }
        map<char, int> roman = {{'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}};
        int total = 0;

        for (int i = 0; i < s.length(); i++) {
            // If the current value is less than the next value, subtract it
            if (i < s.length() - 1 && roman[s[i]] < roman[s[i + 1]]) {
                total += roman[s[i + 1]] - roman[s[i]];
                i++; // Skip the next character
            } else {
                total += roman[s[i]];
            }
        }

    return total;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const string& s) {
        if (s.size() < 1 || s.size() > 15) return false;
        
        return true;
    }
};

File: strings/07_roman_to_int/roman_to_int.py
class Solution:
    def roman_to_int(self, s: str) -> int:
        if not self._check_constraints(s):
            raise ValueError("Constraints violated")

        roman = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        total = 0

        for i in range(len(s)):
            # If the current value is less than the next value, subtract it
            if i < len(s) - 1 and roman[s[i]] < roman[s[i + 1]]:
                total -= roman[s[i]]
            else:
                total += roman[s[i]]

        if (total < 1) or (total > 3999):
            raise ValueError("Constraints violated")

        return total

    def _check_constraints(self, s: str) -> bool:
        if len(s) < 1 or len(s) > 15:
            return False
        roman = set(['I', 'V', 'X', 'L', 'C', 'D', 'M'])
        for ch in s:
            if ch not in roman:
                return False

        return True
